README.md
<h3 align="center">
  babel-preset-latest-node
</h3>

<p align="center">
  babel preset latest for node 14 (Current), node 12 (Active LTS) node 10 (Maintenance LTS)
</p>

<p align="center">
  <a href="https://npmjs.org/package/babel-preset-latest-node"><img src="https://img.shields.io/npm/v/babel-preset-latest-node.svg?style=flat-square"></a>
</p>

# babel 7

Since v2, this package requires `@babel/core@^7.0.0`. If you use babel 6, you can still use the version "1.0.0" of this package. If you want to migrate, you can read the [announcement](https://babeljs.io/blog/2018/08/27/7.0.0) and the [official migration guide](https://babeljs.io/docs/en/v7-migration).

## Alternatives

- [@babel/preset-env](https://www.npmjs.com/package/@babel/preset-env), especially `targets.node`

## Options

- `target`: `10`, `10.13`, `12`, `12.10`, `14` or `'current'` (`process.versions.node`)
- `loose`: Enable “loose” transformations for any plugins in this preset that allow them (Disabled by default).
- `modules` - Enable transformation of ES6 module syntax to another module type (Enabled by default to "commonjs"). Can be false to not transform modules, or "commonjs"
- `es2019` - Enable es2019 features (Enabled by default)
- `es2020` - Enable es2020 features (Enabled by default)
- `shippedProposals` - Enable features in stages but already available in recent node version (Enabled by default)

## [Compatibility Table](http://node.green/)

| Feature                                                                                                         | Node 10 | Node 10.13 | Node 12 | Node 14   | Node 14.8 |
| --------------------------------------------------------------------------------------------------------------- | ------- | ---------- | ------- | --------- | --------- |
| <h3>ES2015</h3>                                                                                                 |         |            |         |           |
| [transform-modules-commonjs](https://www.npmjs.com/package/@babel/plugin-transform-modules-commonjs)            | flag    | flag       | flag    | with .mjs | with .mjs |
| <h3>ES2019</h3>                                                                                                 |         |            |         |           |
| [syntax-optional-catch-binding](https://www.npmjs.com/package/@babel/plugin-syntax-optional-catch-binding)      | flag    | yes        | yes     | yes       | yes       |
| [json-strings](https://www.npmjs.com/package/@babel/plugin-proposal-json-strings)                               | yes     | yes        | yes     | yes       | yes       |
| <h3>ES2020</h3>                                                                                                 |         |            |         |           |
| [nullish-coalescing-operator](https://www.npmjs.com/package/@babel/plugin-proposal-nullish-coalescing-operator) | no      | no         | no      | yes       | yes       |
| [optional-chaining](https://www.npmjs.com/package/@babel/plugin-proposal-optional-chaining)                     | no      | no         | no      | yes       | yes       |
| <h3>Shipped Proposals</h3>                                                                                      |         |            |         |           |
| [numeric-separator](https://www.npmjs.com/package/@babel/plugin-proposal-numeric-separator)                     | no      | no         | no      | yes       | yes       |
| [top-level-await](https://www.npmjs.com/package/@babel/plugin-syntax-top-level-await)                           | no      | no         | no      | no        | yes       |

Note that top level await is syntax only and is only available for supported versions.

## Install

```bash
npm install --save-dev @babel/core babel-preset-latest-node
yarn add --dev @babel/core babel-preset-latest-node
```

## Usage

### Via `.babelrc`

**.babelrc**

```json
{
  "presets": ["latest-node"]
}
```

```json
{
  "presets": [["latest-node", { "target": "current" }]]
}
```

### Via CLI

```sh
babel script.js --presets latest-node
```

### Via Node API

```javascript
require('babel-core').transform('code', {
  presets: [require('babel-preset-latest-node')],
});
```

```javascript
require('babel-core').transform('code', {
  presets: [[require('babel-preset-latest-node'), { target: 'current' }]],
});
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [5.5.1](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v5.5.0...v5.5.1) (2021-03-14)


### Bug Fixes

* remove yarn in engines ([5b17c78](https://github.com/christophehurpeau/babel-preset-latest-node/commit/5b17c7847085643a8406ea8c731c14e401c44a46))

## [5.5.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v5.4.0...v5.5.0) (2021-03-14)


### Features

* update babel deps ([0f4b12a](https://github.com/christophehurpeau/babel-preset-latest-node/commit/0f4b12ad90e3de962c0a5acf4956ca507c109acd))


### Bug Fixes

* use proposal-numeric-separator for node < 14 ([d50e121](https://github.com/christophehurpeau/babel-preset-latest-node/commit/d50e1212adeb88be5ce1203a1c057148c4178a3f))

## [5.4.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v5.2.0...v5.4.0) (2021-01-17)

### Features

- update babel deps ([a732b74](https://github.com/christophehurpeau/babel-preset-latest-node/commit/a732b74df2f0b1770437714bfb58334316fbe2e7))

## [5.3.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v5.2.0...v5.3.0) (2021-01-17)

### Features

- update babel deps ([a732b74](https://github.com/christophehurpeau/babel-preset-latest-node/commit/a732b74df2f0b1770437714bfb58334316fbe2e7))

## [5.2.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v5.1.1...v5.2.0) (2020-12-05)

### Features

- add @babel/plugin-syntax-top-level-await for node 14.8 and higher ([47c8c88](https://github.com/christophehurpeau/babel-preset-latest-node/commit/47c8c886626aff962ba27e9153d279d702e294c0))

### [5.1.1](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v5.1.0...v5.1.1) (2020-11-13)

### Bug Fixes

- remove yarn in engines ([83318fe](https://github.com/christophehurpeau/babel-preset-latest-node/commit/83318fe0b52f3ec606bb8adc73ab4be8ef2a1354))

## [5.1.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v5.0.0...v5.1.0) (2020-11-13)

### Features

- accept target 12.10 ([ed16358](https://github.com/christophehurpeau/babel-preset-latest-node/commit/ed16358c128e7df0534afca64b17d93acc268e4b))
- set sideEffects in package.json ([9464031](https://github.com/christophehurpeau/babel-preset-latest-node/commit/9464031686d0432d40ffa8f7f4e3ecba5bdb731c))

## [5.0.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v4.1.0...v5.0.0) (2020-09-19)

### ⚠ BREAKING CHANGES

- drop target 11 and 13
- requires latest @babel/core

### Features

- add node 14, es2020 and numeric-separator for shipped proposals ([a6437f3](https://github.com/christophehurpeau/babel-preset-latest-node/commit/a6437f354cb22162f4b861249f224c63dbee7ccb))
- drop target 11 and 13 ([5adddfd](https://github.com/christophehurpeau/babel-preset-latest-node/commit/5adddfdbc464169d7acbea8afb93488961acf0f4))

## [4.0.1](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v4.0.0...v4.0.1) (2019-12-14)

### Bug Fixes

- require default from latest plugins ([30708ae](https://github.com/christophehurpeau/babel-preset-latest-node/commit/30708aea524cdd031857e767bd7ec3fa076cb619))

# [4.0.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v3.2.1...v4.0.0) (2019-12-13)

### Features

- add nullish coalescing operator and optional chaining ([f9e89f1](https://github.com/christophehurpeau/babel-preset-latest-node/commit/f9e89f1a1a1a55609c567fae0765ffcdad648471))
- add support for node 12 ([212b1b9](https://github.com/christophehurpeau/babel-preset-latest-node/commit/212b1b9f7a6f11c00df9d2035e80c914d3077964))
- drop node 6 and node 8 ([5974020](https://github.com/christophehurpeau/babel-preset-latest-node/commit/5974020f29ff18765904cffe5a26b41c8f15600f))
- update dependencies ([a50d541](https://github.com/christophehurpeau/babel-preset-latest-node/commit/a50d5410cc37d375fedd649835b4185bf4ead075))

### BREAKING CHANGES

- node 6 and node 8 no longer supported, removed options es2016, es2017, es2018 and esnext (replaced by shippedProposals)

## [3.2.1](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v3.2.0...v3.2.1) (2019-05-09)

### Bug Fixes

- check option es2019 ([b959ffd](https://github.com/christophehurpeau/babel-preset-latest-node/commit/b959ffd26ad1545a228d92468751eb6765ee3831))

# [3.2.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v3.1.1...v3.2.0) (2019-05-09)

### Features

- add es2019 option and plugin-proposal-json-strings in shippedProposals ([dfc2534](https://github.com/christophehurpeau/babel-preset-latest-node/commit/dfc25344d61bbe565c52e5f2faa78854514bfea0))
- support node 12 ([38037b9](https://github.com/christophehurpeau/babel-preset-latest-node/commit/38037b91d81399f816fbe88d6f977ea3612697ea))
- update dependencies ([b0d6da8](https://github.com/christophehurpeau/babel-preset-latest-node/commit/b0d6da8c4015e66f45baee2c13436c3e7aac7024))

## [3.1.1](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v3.1.0...v3.1.1) (2019-03-12)

### Bug Fixes

- make target option optional ([76fdc33](https://github.com/christophehurpeau/babel-preset-latest-node/commit/76fdc3333d5853db83e386a67b060dd59eb10885)), closes [#60](https://github.com/christophehurpeau/babel-preset-latest-node/issues/60)

# [3.1.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v3.0.1...v3.1.0) (2019-03-09)

### Features

- update deps ([0baad94](https://github.com/christophehurpeau/babel-preset-latest-node/commit/0baad9473e02de7adda524fbf5113fe760a4f9b2))

## [3.0.1](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v3.0.0...v3.0.1) (2018-11-23)

# [3.0.0](https://github.com/christophehurpeau/babel-preset-latest-node/compare/v2.1.0...v3.0.0) (2018-11-23)

### Features


lib/index.js
'use strict';

const validTargetOption = [
  '10',
  '10.13',
  '12',
  '12.10',
  '14',
  '14.8',
  'current',
];

const checkTargetOption = (targetOption) => {
  if (targetOption && !validTargetOption.includes(targetOption)) {
    throw new Error(
      `Preset latest-node 'target' option must one of ${validTargetOption.join(
        ', ',
      )}.`,
    );
  }
};

const checkDeprecatedOptions = (opts) => {
  if (opts.esnext !== undefined) {
    throw new Error(
      'Preset latest-node you should replace esnext by shippedProposals.',
    );
  }
};

const checkBooleanOptions = (opts) => {
  ['loose', 'es2019', 'shippedProposals'].forEach((optionName) => {
    if (
      opts[optionName] !== undefined &&
      typeof opts[optionName] !== 'boolean'
    ) {
      throw new Error(
        `Preset latest-node '${optionName}' option must be a boolean.`,
      );
    }
  });
};

const checkModulesValue = (modules) => {
  if (modules !== false && modules !== 'commonjs') {
    throw new Error(
      "Preset latest-node 'modules' option must be 'false' to indicate no modules\n" +
        "or 'commonjs' (default)",
    );
  }
};
module.exports = function (context, opts = {}) {
  checkBooleanOptions(opts);
  checkDeprecatedOptions(opts);

  const targetOption = !opts.target ? undefined : String(opts.target);
  checkTargetOption(targetOption);

  const target =
    targetOption && targetOption !== 'current'
      ? targetOption
      : process.versions.node;
  const splittedTarget = target.split('.');

  const major = Number(splittedTarget[0]);
  const minor = Number(splittedTarget[1]) || 0;

  const loose = opts.loose !== undefined ? opts.loose : false;
  const modules = opts.modules !== undefined ? opts.modules : 'commonjs';
  const es2019 = opts.es2019 !== undefined ? opts.es2019 : true;
  const es2020 = opts.es2020 !== undefined ? opts.es2020 : true;
  const shippedProposals =
    opts.shippedProposals !== undefined ? opts.shippedProposals : true;

  checkModulesValue(modules);

  const optsLoose = { loose };
  const infNode10dot13 = major < 10 || (major === 10 && minor < 13);
  const infNode14 = major < 14;
  const supOrEqNode14dot8 = major > 15 || (major === 14 && minor >= 8);

  return {
    plugins: [
      /* es2015 */
      modules === 'commonjs' && [
        require('@babel/plugin-transform-modules-commonjs'),
        optsLoose,
      ],

      /* es2019 */
      es2019 && infNode10dot13
        ? require('@babel/plugin-proposal-optional-catch-binding')
        : undefined,

      /* es2020 */
      es2020 &&
        infNode14 &&
        require('@babel/plugin-proposal-nullish-coalescing-operator').default,
      es2020 &&
        infNode14 &&
        require('@babel/plugin-proposal-optional-chaining').default,

      /* shippedProposals */
      shippedProposals &&
        infNode14 &&
        require('@babel/plugin-proposal-numeric-separator').default,
      shippedProposals &&
        supOrEqNode14dot8 &&
        require('@babel/plugin-syntax-top-level-await').default,
    ].filter(Boolean),
  };
};
